Auto-include types for the jsx import source in the new jsx transforms

weswigham opened the pull request:
Fixes #41118
Fixes #40502

A note on #41118 for @ddprrt - `preact`'s source, in that case, is currently (re)exporting a `JSX` namespace under each factory function (multiple factory functions are used within a compilation). That's not quite what I'm looking for, for local JSX namespaces on implicit jsx runtime imports with this PR - a single top-level `JSX` export, alongside all those factory functions, is instead what I want to find. This should simplify the file a bit, as a single `export {JSXInternal as JSX} from "../../"` should suffice.

weswigham commented:
@typescript-bot pack this
@typescript-bot user test this
@typescript-bot perf test this
@typescript-bot test this
@typescript-bot run dt

ddprrt commented:
Oh cool, thank you! Is this something we should carry over to Preact? (I assume it's similar to tests/cases/compiler/jsxNamespaceImplicitImportJSXNamespace.tsx) -- I can work on a PR and maybe @marvinhagemeister is willing to guide me 😄

weswigham commented:
Yeah, probably! For the implicitly added imports, the full factory is `jsx_import_source.jsx` or `jsx_import_source.jsxs` or `ksx_import_source.jsxDEV`, so the place we look for the JSX namespace within is the `jsx_import_source` itself. The change to preact is pretty small - adding `export {JSXInternal as JSX};` to the bottom is sufficient, but it can be cleaned up more from there (as you _probably_ don't need all the other reexports).

sandersn reviewed (approved):
I looked at the checker changes and they seem reasonable. I don't know enough about the program.ts change, but the comment's explanation makes sense to me.

orta reviewed (approved):
Mostly looked over the test cases, and this works like I'd expect 👍🏻

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -402,8 +402,10 @@ namespace ts {
     export function getAutomaticTypeDirectiveNames(options: CompilerOptions, host: ModuleResolutionHost): string[] {
         // Use explicit type list from tsconfig.json
+        // jsxImportSource, if present and in use, creates implicit imports
+        const implicitImport = getJSXRuntimeImport(getJSXImplicitImportBase(options), options);
```
I am not sure we should be adding anything implicit if `types` is specified.  Thats exactly why that option exists.

weswigham replied:
I add it to the `types` list because I don't think there's a great reason to make people repeat the same (or nearly the same) module name for both the `jsxImportSource` and the `types` flags - it just seems to me to make sense to include it in the `types` list, if present.

sheetalkamat replied:
Question for @DanielRosenwasser then. If we are always going to implicitly add this we should process this separately in processTypeReferenceDirective in program.ts instead of modifying list here. 

I think apart from incremental we should add test where the file is added and later goes missing for this implicit import with and without types specified because we handle them specifically for watching in resolutionCache

DanielRosenwasser replied:
Isn't this sort of a similar concept to how `tslib` is added with `importHelpers`? You've got an implicit import injected and we do some resolution to see if it's vaguely compatible, right?

@RyanCavanaugh and @andrewbranch might want to weigh in here as well.

weswigham replied:
Yeah, I guess so - primary difference being we actually end up pulling the types from the module for checking things.

DanielRosenwasser replied:
Looks like we might do the same for tslib?

```ts
function checkExternalEmitHelpers(location: Node, helpers: ExternalEmitHelpers) {
    if ((requestedExternalEmitHelpers & helpers) !== helpers && compilerOptions.importHelpers) {
        const sourceFile = getSourceFileOfNode(location);
        if (isEffectiveExternalModule(sourceFile, compilerOptions) && !(location.flags & NodeFlags.Ambient)) {
            const helpersModule = resolveHelpersModule(sourceFile, location);
            if (helpersModule !== unknownSymbol) {
                const uncheckedHelpers = helpers & ~requestedExternalEmitHelpers;
                for (let helper = ExternalEmitHelpers.FirstEmitHelper; helper <= ExternalEmitHelpers.LastEmitHelper; helper <<= 1) {
                    if (uncheckedHelpers & helper) {
                        const name = getHelperName(helper);
                        const symbol = getSymbol(helpersModule.exports!, escapeLeadingUnderscores(name), SymbolFlags.Value);
                        if (!symbol) {
                            error(location, Diagnostics.This_syntax_requires_an_imported_helper_named_1_which_does_not_exist_in_0_Consider_upgrading_your_version_of_0, externalHelpersModuleNameText, name);
                        }
                    }
                }
            }
            requestedExternalEmitHelpers |= helpers;
        }
    }
}
```

weswigham replied:
Hm, for `importHelpers` we make a synthetic `import` within `collectExternalModuleReferences` - I could do that; hopefully it doesn't have other effects?

DanielRosenwasser replied:
Not sure - @rbuckton might know more here too.

weswigham replied:
I've changed the implementation to use the same synthetic-import method that we use for `importHelpers`.

sheetalkamat commented on src/compiler/program.ts:
```diff
@@ -3882,14 +3887,19 @@ namespace ts {
             }
             // Do nothing if it's an Identifier; we don't need to do module resolution for `declare global`.
         }
+        const jsxRuntimeImport = getJSXRuntimeImport(getJSXImplicitImportBase(options, file), options);
```
You need test for incremental build and need to be handled in builder to store the file reference with this import.

sheetalkamat commented on src/compiler/utilities.ts:
```diff
@@ -6009,8 +6009,8 @@ namespace ts {
-    export function getJSXImplicitImportBase(compilerOptions: CompilerOptions, file: SourceFile): string | undefined {
-        const jsxImportSourcePragmas = file.pragmas.get("jsximportsource");
+    export function getJSXImplicitImportBase(compilerOptions: CompilerOptions, file?: SourceFile): string | undefined {
+        const jsxImportSourcePragmas = file?.pragmas.get("jsximportsource");
```
You would also need to mark `jsxImportSource` as affects module resolution  and `jsx` as well ? can you add test cases to make sure update in that reflects change in program.

weswigham commented:
@sheetalkamat I've added the affixes to the compiler options to make them affect incremental emit, and added an incremental emit test.

sheetalkamat commented on src/compiler/program.ts:
```diff
@@ -2216,16 +2225,18 @@ namespace ts {
+                const jsxImport = getJSXRuntimeImport(getJSXImplicitImportBase(options, file), options);
+                if (jsxImport) {
+                    // synthesize `import "base/jsx-runtime"` declaration
+                    imports ||= [];
+                    imports.push(createSyntheticImport(jsxImport, file));
```
Like this!

sheetalkamat replied:
```suggestion
                    (imports ||= []).push(createSyntheticImport(jsxImport, file));
```

sheetalkamat commented on src/testRunner/unittests/tscWatch/incremental.ts:
```diff
@@ -276,5 +276,33 @@ export interface A {
+                { path: `${project}/index.tsx`, content: `export const App = () => <div propA={true}></div>;` },
+                { path: configFile.path, content: JSON.stringify({ compilerOptions: jsxImportSourceOptions }) }
+            ],
+            modifyFs: host => host.writeFile(configFile.path, JSON.stringify({ compilerOptions: { ...jsxImportSourceOptions, jsxImportSource: "preact" } }))
```
Can you add another case hen preact\*.d.ts file is missing and added.
And another when it is present and as incremental change goes missing.

DanielRosenwasser commented:
btw I'm assuming I should wait on this for 4.1 RC - is that reasonable?

weswigham commented:
Yeah, I'd rather this get into the RC than go out in 4.1 post-RC, since it affects the usability of the `react-jsx` transform quite a bit.

sheetalkamat commented:
@weswigham the change now looks good and adding tests for https://github.com/microsoft/TypeScript/pull/41330#discussion_r515420969 should make it ready to merge

DanielRosenwasser commented:
If we can get the tests in and passing as above, I'll merge this in and kick off the appropriate builds.

sheetalkamat reviewed (approved):
discussed offline with @weswigham that there is existing bug with incremental build with missing react files being available later which he will work on later  and is not related to this change.

weswigham commented:
(specifically there's an issue with synthetic `imports` like the one used here and in `importHelpers` output that causes incremental emit to not pick up the dependency relationship between the files)

DanielRosenwasser commented:
So what's the idea? Cherry-pick that to 4.1 but leave it out of 4.1 RC?

weswigham commented:
Well, I got a fix for that underlying issue together pretty quickly, since the issue was actually pretty trivial - our public API sanitity checks forbid answering questions on synthetic declarations like those _we'd_ made - https://github.com/microsoft/TypeScript/pull/41346 is out. Up to you. If you can get a review from people you're happy with in a timeframe you're OK with, it can make it into the RC, otherwise saving the fix for post-RC cherry-pick is fine.

Andarist commented:
@weswigham is there a way to make this work:
```ts
import { createElement } from 'react'
import { EmotionJSX } from './jsx-namespace'

export const jsx: typeof createElement
export namespace jsx {
  export import JSX = EmotionJSX
}
```
It complains about "Cannot redeclare block-scoped variable 'jsx'.ts(2451)" even though [this works](https://github.com/emotion-js/emotion/blob/0d3420034561908776024b2e3a3bfc5907804350/packages/react/types/index.d.ts#L109-L132) so it's clearly that `export import` syntax triggering this error here.

I've tried to make it work for several hours already and can't do it so far.

weswigham commented:
It's a limitation of our alias-following logic in the binder, if I'm reading it right. If the `EmotionJSX` namespace was defined in the same file, it'd probably work, because then the binder would be able to know that `EmotionJSX` only refers to a namespace containing only types (by following the aliases within the file), which then means `jsx` contains only types, which is consequently safe to merge with `const jsx`. Across files, we don't know what `EmotionJSX` consists of until much later during checking, so we conservatively disallow the merge (since it might have some value member which could conflict with some value in the `const`)... at least that's my read on it.
Naively, I'd hope that
```ts
import type { EmotionJSX } from './jsx-namespace'
```
would cause it to work, but I don't know if we adjusted the alias following logic in the _binder_ to allow that to work when we added `import type`. (Consequently, if _that_ doesn't work, I think an isolated repro of the merge using `import type` and a new issue would be welcome)

Andarist commented:
@weswigham thanks for the tip! I'm afraid it doesn't work though. It also gives me one additional error and that is:
> An import alias cannot reference a declaration that was imported using 'import type'

If I find the time I will create a repro case for this and report the appropriate issue. In the meantime I've filled a followup PR to this one: https://github.com/microsoft/TypeScript/pull/41476 , please take a look at it when you have time, would be great to fix the issue mentioned there before the final release.
